### DISCUSSION TITLE
How to output quantum states for each time step using an optimized control pulse?

### DISCUSSION THREAD

**Discussion by Leeseok-628** [qucontrol/krotov#113] [Q&A]
After obtaining an optimized control pulse solution using Krotov's package, I would like to print out density matrices for each time step. Initially, I attempted to accomplish this using `qutip.mesolve`, but found that the resulting solutions were different (e.g., see the code below). I then attempted to use the propagator function (after reading through [this notebook](https://nbviewer.org/gist/goerz/21e46ea7b45c9514e460007de14419bd/Krotov_time_discretization.ipynb)), but still encountered discrepancies.

Therefore, my inquiry is: What is the correct way to output quantum states for each time step, given a list of optimized control pulses?

To make my question clearer, below is an example of my code that failed. What I want is to make the fidelity to $1$. However, both the `qutip.mesolve` and `propagator` function methods gave fidelity values below 1. What could be the issue with my code?

<img width="987" alt="Screen Shot 2023-03-24 at 10 51 39 AM" src="https://user-images.githubusercontent.com/55151494/227590065-eac43578-da8b-4531-b93d-9cfba51df7e4.png">

**Comment by goerz** [comment]
Why should the propagation return a fidelity of 1? Optimal control is generally asymptotic: you can get close to 1, but you won't reach 1. In this case, it looks like you stopped the optimization when you reached 99.6%. The discrepancy between `mesolve` and `propagate` is due to the difference in discretization. This discrepancy should get smaller if you make your time step smaller, but beyond that, what you're seeing is expected behavior.

**Reply by Leeseok-628** [reply]
Thank you for your response, Goerz. I am comparing the fidelity between two quantum states, one generated by the optimal control pulse obtained from Krotov's method using `opt_result.state`, and the other generated by using the same optimized control pulse with either the `propagator` or `mesolve` function using `opt_result.optimized_controls`. Since both states are generated by the same control pulse, I would expect them to be identical. However, I have found that they are not the same as shown in the code above, which leads me to suspect that I may have made some mistakes when using the `propagator` or `mesolve` function, such as using different `tlist` values. I'm not sure what I did wrong though; for example in the first block of the code, I put the same optimal control pulse, the same `tlist` that I used earlier for my optimization, and the same Lindbladian. So my question is how to generate the same state as `opt_result.state` by using either the `propagator` or `mesolve` function using `opt_result.optimized_controls`.

**Reply by goerz** [reply]
Oh, I see… Sorry for the misunderstanding! You're right, the `propagate` method should reproduce the state in `opt_result.states`, assuming you're using the same propagator with the same settings as was used in the optimization. It's not immediately obvious to me what's wrong with your code.

**Reply by goerz** [reply]
If you can create a minimal working example that demonstrates this behavior, you might want to open an issue
